### PR TITLE
Spec fixes

### PR DESCRIPTION
**Pull request by tom-wa**

### CODE REVIEW

File: src/plugins/conditionals/testmod_conditionals.c [hunk 1]
```diff
@@ -201,25 +201,67 @@ static void test_assignElse ()
 	keyDel (parentKey);
 	PLUGIN_CLOSE ();
 }
-int main (int argc, char ** argv)
+
+static void test_assignKeyThen()
+{
+	Key *parentKey = keyNew("user/tests/conditionals", KEY_VALUE, "", KEY_END);
```
Comment (markus2330) [inline src/plugins/conditionals/testmod_conditionals.c]: None of the tests use a null-key?

**Comment by markus2330** [comment]
Thank you! Can you please apply `clang-format`?

**Comment by markus2330** [comment]
There is now a memleak in isAssign, see: http://build.libelektra.org:8080/job/elektra-mergerequests/839/artifact/build/Testing/20160223-0353/DynamicAnalysis.xml

File: src/plugins/conditionals/testmod_conditionals.c (hunk 1 above)
Comment (tom-wa) [inline src/plugins/conditionals/testmod_conditionals.c]: I'll add a shell testscript for it, the issue was related to the spec plugin, not the conditionals plugin (ok, i forgot the postgetstorage placement in the contract, but that wasn't exactly the issue)

**Comment by markus2330** [comment]
Thank you! There are many whitespace issues and removal of `{}`. You should easily see them in the PR. Whitespace issues should be resolved with a clang-format.

File: src/plugins/conditionals/conditionals.c [hunk 2]
```diff
@@ -325,32 +298,44 @@ static int parseSingleCondition (const char * condition, KeySet * ks, Key * pare
...
 		++firstPtr;
-	}
 	while (isspace (*lastPtr))
-	{
 		--lastPtr;
-	}
 	if (*firstPtr != '\'' || *lastPtr != '\'')
 	{
-		return NULL;
+		if (lastPtr <= firstPtr)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: Can you document this a bit, please?

File: src/plugins/conditionals/conditionals.c [hunk 3]
```diff
@@ -505,14 +490,59 @@ static int parseConditionString (const Key * meta, Key * parentKey, Key * key, K
...
 	if (elseexpr)
-	{
 		elektraFree (elseexpr);
-	}
 	regfree (&regex);
 	ksDel (ks);
 	return ret;
 }
 
+static int doCheck (Key * meta, Key * parentKey, Key * key, KeySet * ks, Operation op)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: name, too generic, which check is done? Please comment.

File: src/plugins/conditionals/conditionals.c [hunk 4]
```diff
@@ -530,101 +560,144 @@ int elektraConditionalsGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned EL
...
-			else if (result == 0)
-			{
-				ret |= -1;
-			}
-			else if (result == 1)
-			{
-				ret |= 1;
+				int result;
+				result = doCheck (conditionMeta, parentKey, cur, returned, CONDITION);
+				if (result == -3)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: Why -3? Why not use an enum?

File: src/plugins/spec/spec.c [hunk 5]
```diff
@@ -578,7 +589,298 @@ static int handleErrors (Key * parentKey, KeySet * ks, Key * key, Key * specKey,
 	return ret;
 }
 
-static void copyMeta (Key * key, Key * specKey)
+static void matchedKeyCopyMeta (Key * key, Key * specKey, Key * parentKey, const char * name)
```
Comment (markus2330) [inline src/plugins/spec/spec.c]: Some documentation should be here, too.

**Comment by markus2330** [comment]
Please ping me if ready!

**Comment by tom-wa** [comment]
this PR also fixes #521 and #522.
partitional fix for #523 (only for set, still trying to figure out the best solution for get)

File: src/plugins/spec/spec.c (hunk 5 above)
Comment (markus2330) [inline src/plugins/spec/spec.c]: Docu is still missing and this function is way too long. Please split it or at least comment the most important pieces.

File: src/plugins/conditionals/README.md [hunk 6]
```diff
@@ -4,9 +4,9 @@
 - infos/provides =
 - infos/needs =
 - infos/recommends =
-- infos/placements = presetstorage
+- infos/placements = postgetstorage presetstorage
 - infos/status = maintained unittest nodep libc preview
-- infos/metadata = check/condition assign/condition
+- infos/metadata = check/condition assign/condition condition/validsuffix
```
Comment (markus2330) [inline src/plugins/conditionals/README.md]: No docu on validsuffix.

**Comment by markus2330** [comment]
Thank you!

The PR is too diverse, please prefer in future to split it up in smaller parts. Some pieces would be perfectly acceptable, but others (especially the long undocumented function in spec) are not.

File: src/plugins/conditionals/conditionals.c [hunk 7]
```diff
@@ -76,23 +116,29 @@ static int compareStrings (const char * s1, const char * s2)
...
+	{
+		retval = 0;
+	}
+	else if ((ret = isNumber (s1, suffixList)) && (ret2 = isNumber (s2, suffixList)))
 	{
 		if (ret == 2 || ret2 == 2)
 		{
 			result = (fabs (atof (s1) - atof (s2)));
 			if (result < EPSILON)
-			{
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: Please keep such `{}`

File: src/plugins/conditionals/conditionals.c [hunk 8]
```diff
@@ -107,56 +153,62 @@ static int compareStrings (const char * s1, const char * s2)
...
-		snprintf (lookupName, len, "%s/%s", keyName (parentKey), rightSide);
-		key = ksLookupByName (ks, lookupName, 0);
-		if (!key)
-		{
-			ELEKTRA_SET_ERRORF (133, parentKey, "Key %s not found but is required for the evaluation of %s", lookupName,
-					    condition);
-			result = -2;
-			goto Cleanup;
+			char * endPos = strchr (rightSide + 1, '\'');
+			if (!endPos)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: here should be docu, too.

File: src/plugins/conditionals/conditionals.c [hunk 9]
```diff
@@ -504,15 +638,43 @@ static int parseConditionString (const Key * meta, Key * parentKey, Key * key, K
...
-	{
-		elektraFree (elseexpr);
-	}
+	if (elseexpr) elektraFree (elseexpr);
 	regfree (&regex);
 	ksDel (ks);
 	return ret;
 }
 
+static int parse (const Key * meta, const Key * suffixList, Key * parentKey, Key * key, KeySet * ks, Operation op)
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: name too vague.

File: src/plugins/conditionals/conditionals.c [hunk 10]
```diff
@@ -168,140 +220,155 @@ static int evalCondition (const char * leftSide, Comparator cmpOp, const char *
...
 }
 
 
-static int parseSingleCondition (const char * condition, KeySet * ks, Key * parentKey)
+static int parseSingleCondition (const char * condition, const Key * suffixList, KeySet * ks, Key * parentKey)
 {
 	Comparator cmpOp;
 	char * opStr;
 	if ((opStr = strstr (condition, "==")))
-	{
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: please keep the `{}`. I think you can easily refactor this part to a condition2cmpOp function.

File: src/plugins/conditionals/conditionals.c [hunk 11]
```diff
@@ -491,8 +626,7 @@ static int parseConditionString (const Key * meta, Key * parentKey, Key * key, K
 		}
 		else
 		{
-			ELEKTRA_SET_ERRORF (135, parentKey, "Validation of %s failed. (%s failed) ", conditionString, condition);
-			ret = 1;
+			ret = -3;
```
Comment (markus2330) [inline src/plugins/conditionals/conditionals.c]: please avoid these magic numbers and prefer to use an enum instead.

File: src/plugins/enum/enum.c [hunk 12]
```diff
@@ -104,10 +101,5 @@ int elektraEnumSet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UN
 
 Plugin * ELEKTRA_PLUGIN_EXPORT (enum)
 {
-	// clang-format off
```
Comment (markus2330) [inline src/plugins/enum/enum.c]: do not remove clang-format here, its not well-formated then.

File: src/plugins/ini/contract.h [hunk 13]
```diff
@@ -12,29 +12,16 @@
...
-			keyNew ("system/elektra/modules/ini/exports/close",
-					KEY_FUNC, elektraIniClose,
-					KEY_END),
-	#include "readme_ini.c"
-			keyNew ("system/elektra/modules/ini/infos/version",
-					KEY_VALUE, PLUGINVERSION,
-					KEY_END),
-			KS_END);
-	// clang-format on
+	// @formatter:off
```
Comment (markus2330) [inline src/plugins/ini/contract.h]: it seems like the merge did not work well here.

File: src/plugins/ini/ini.c [hunk 14]
```diff
@@ -1704,12 +1715,6 @@ int elektraIniSet (Plugin * handle, KeySet * returned, Key * parentKey)
...
 Plugin * ELEKTRA_PLUGIN_EXPORT (ini)
 {
-	// clang-format off
-	return elektraPluginExport("ini",
-			ELEKTRA_PLUGIN_OPEN, &elektraIniOpen,
-			ELEKTRA_PLUGIN_CLOSE, &elektraIniClose,
-			ELEKTRA_PLUGIN_GET,	&elektraIniGet,
-			ELEKTRA_PLUGIN_SET,	&elektraIniSet,
-			ELEKTRA_PLUGIN_END);
+	return elektraPluginExport ("ini", ELEKTRA_PLUGIN_OPEN, &elektraIniOpen, ELEKTRA_PLUGIN_CLOSE, &elektraIniClose, ELEKTRA_PLUGIN_GET,
```
Comment (markus2330) [inline src/plugins/ini/ini.c]: again clang-format.

File: src/plugins/ini/inih-r29/inih.c [hunk 15]
```diff
@@ -26,9 +26,7 @@ static char * rstrip (char * s)
 {
 	char * p = s + strlen (s);
 	while (p > s && isspace ((unsigned char)(*--p)))
-	{
```
Comment (markus2330) [inline src/plugins/ini/inih-r29/inih.c]: Merge problem?

File: src/plugins/spec/spec.c [hunk 16]
```diff
@@ -578,7 +589,354 @@ static int handleErrors (Key * parentKey, KeySet * ks, Key * key, Key * specKey,
 	return ret;
 }
 
-static void copyMeta (Key * key, Key * specKey)
+
+// Naive approach to replace wildcards in meta string with the actual values
```
Comment (markus2330) [inline src/plugins/spec/spec.c]: Please use doxygen comments here. A bit more explaination would be helpful. Is this the work-in-progress we discussed?

File: src/plugins/spec/spec.c [hunk 17]
```diff
@@ -578,7 +589,354 @@ static int handleErrors (Key * parentKey, KeySet * ks, Key * key, Key * specKey,
...
+				char * ptrCopy = ptr;
+
+				// check if the key name part containing the wildcard matches the globbing key
+
+				// check starting at the wildcard until the start of the name part.
+				while (!noMatch && ptrCopy > relativeName && segmentStart > metaString)
+				{
+					if (*ptrCopy == '?')
+					{
+						while (*segmentStart != '/' && segmentStart > metaString && ptrCopy > relativeName &&
```
Comment (markus2330) [inline src/plugins/spec/spec.c]: what is happening here? Looks very similar to code block below

**Comment by markus2330** [comment]
Did you forget to ping? Its much better now!

**Comment by tom-wa** [comment]
@markus2330 yeah sorry, went to something else and forgot it.

File: src/plugins/spec/spec.c (hunk 17 above)
Comment (tom-wa) [inline src/plugins/spec/spec.c]: it's pretty much the same. the first block searches from the wildcard position to the beginning of the keyname part and check if it matches the part from the globbing string, the second block, only if the first succeeded, from the wildcard position to the name part end

**Comment by markus2330** [comment]
Thank you, nice fixes! I hope, however, we find a better solution for wildcards in metadata before 0.8.16.

**Pull request merged**